grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -0,0 +1,185 @@
+# Lint as: python3
+from absl import app, flags
+import json
+import os
+from typing import Any, cast, Dict, List, Text, Tuple
+
+from fleetspeak.src.server.proto.fleetspeak_server import resource_pb2
+
+from grr_response_core import config
+from grr_response_core.config import server as config_server
+from grr_response_core.config import contexts
+from grr_response_server import fleetspeak_connector
+from grr_response_server import fleetspeak_utils
+from grr_response_server import server_startup
+
+from werkzeug import exceptions as werkzeug_exceptions
+from werkzeug import routing as werkzeug_routing
+from werkzeug import serving as werkzeug_serving
+from werkzeug import wrappers as werkzeug_wrappers
+from werkzeug.wrappers import json as werkzeug_wrappers_json
+
+
+AVAILABLE_METRICS = [
+    "mean_user_cpu_rate",
+    "max_user_cpu_rate",
+    "mean_system_cpu_rate",
+    "max_system_cpu_rate",
+    "mean_resident_memory_mib",
+    "max_resident_memory_mib",
+]
+
+JSON_MIME_TYPE = "application/json"
+
+flags.DEFINE_bool(
+    "version",
+    default=False,
+    allow_override=True,
+    help="Print the GRR console version number and exit immediately.")
+
+
+class JSONRequest(werkzeug_wrappers_json.JSONMixin, werkzeug_wrappers.Request):
+  pass
+
+
+class JSONResponse(werkzeug_wrappers_json.JSONMixin, werkzeug_wrappers.Response):
+  pass
+
+
+class Grrafana(object):
+  """GRRafana HTTP server instance.
+  
+  A full description of all endpoints implemented within this HTTP
+  server can be found in https://github.com/simPod/grafana-json-datasource#api."""
+
+  def __init__(self) -> None:
+    """Constructor."""
+    self.url_map = werkzeug_routing.Map([
+        werkzeug_routing.Rule("/", endpoint="Root", methods=["GET"]),
+        werkzeug_routing.Rule("/search", endpoint="Search", methods=["POST"]),
+        werkzeug_routing.Rule("/query", endpoint="Query", methods=["POST"]),
+        werkzeug_routing.Rule("/annotations", endpoint="Annotations", methods=["POST"]),
+    ])
+
+  def DispatchRequest(self, request: JSONRequest) -> JSONResponse:
+    """Maps requests to different methods."""
+    adapter = self.url_map.bind_to_environ(request.environ)
+    try:
+      endpoint, values = adapter.match()
+      return getattr(self, "On" + endpoint)(request, **values)
+    except werkzeug_exceptions.HTTPException as e:
+      return e
+
+  def WsgiApp(self, environ, start_response) -> werkzeug_wrappers.Response:
+    request = JSONRequest(environ)
+    response = self.DispatchRequest(request)
+    return response(environ, start_response)
+
+  def __call__(self, environ, start_response):
+    return self.WsgiApp(environ, start_response)
+
+  def OnRoot(self, request: JSONRequest) -> JSONResponse:
+    """Returns OK message to database connection check."""
+    return JSONResponse()
+
+  def OnSearch(self, request: JSONRequest) -> JSONResponse:
+    """Depending on the type of request Grafana is issuing, this method returns either
```

Comment:
Per [pystyle](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods), docstrings should have a descriptive one-liner, and then after a blank line can contain more information. The idea is that you can roughly understand what a function does by reading one line only.